docs/architecture/adr-022-transaction-handling.md

```diff
@@ -0,0 +1,50 @@
+# ADR 022: Reliable Transaction Handling
+
+## Changelog
+
+- 2023/08/28: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Nonces are a crucial component in transactions for replay protection; preventing users from having their transaction executed multiple times. Each account has a monotonically increasing number associated with it: the nonce (also referred as the `sequence` in the Cosmos SDK). If the last executed nonce is `n`, then the next transaction signed by that account must have a nonce of `n + 1`. This, however, comes with a few downsides. If a user submits 3 transactions in a row with nonces 6, 7, 8 and while propogating throughout the network they are received by the next proposer as 8, 7, 6. Transactions with nonces 7 and 8 will be rejected and only 6 will make it into the mempool to be proposed in the next block. Furthermore, a local hash-based cache may prevent 7 or 8 from every making it into the mempool. Secondly, if the transaction with nonce 6 happened to have too low priority it would block all other transactions. A user would not be able to overwrite the transaction by submitting the same one yet with a higher fee.
```

Comment:
[typo] every => ever

```suggestion
Nonces are a crucial component in transactions for replay protection; preventing users from having their transaction executed multiple times. Each account has a monotonically increasing number associated with it: the nonce (also referred as the `sequence` in the Cosmos SDK). If the last executed nonce is `n`, then the next transaction signed by that account must have a nonce of `n + 1`. This, however, comes with a few downsides. If a user submits 3 transactions in a row with nonces 6, 7, 8 and while propogating throughout the network they are received by the next proposer as 8, 7, 6. Transactions with nonces 7 and 8 will be rejected and only 6 will make it into the mempool to be proposed in the next block. Furthermore, a local hash-based cache may prevent 7 or 8 from ever making it into the mempool. Secondly, if the transaction with nonce 6 happened to have too low priority it would block all other transactions. A user would not be able to overwrite the transaction by submitting the same one yet with a higher fee.
```

[question]
> Secondly, if the transaction with nonce 6 happened to have too low priority it would block all other transactions. A user would not be able to overwrite the transaction by submitting the same one yet with a higher fee.

- "too low priority" relative to what? 
- "all other transactions" sent by this user?
- if the user submits another transaction with a higher fee and nonce 6, why doesn't it replace the previous tx with a low fee and nonce 6?

Reply:
> "too low priority" relative to what? 

relative to all other transactions that can make it in a block. 

> if the user submits another transaction with a higher fee and nonce 6, why doesn't it replace the previous tx with a low fee and nonce 6?

CheckTx will reject that transaction because it previous incremented the nonce from the previous transaction (from 5 to 6) so now it expects a nonce of 7